julia emse paper: blog (#1879)
betterscientificsoftware/bssw.io: Articles/Blog/2023-11-27-julia-bugs.md

# Come for Syntax, Stay for Speed, and Understand Bugs in Julia Programs 

**Hero Image:**

 - <img src='../../images/julia-bugs-taxonomy.png' />
 

#### Contributed by [Akond Rahman](https://github.com/akondrahman "Akond Rahman's GitHub Profile")

#### Publication date: November 27, 2023 

> In this blog post, I describe some of our [research findings](https://link.springer.com/article/10.1007/s10664-023-10328-5) related to bugs that occur in Julia programs, which are becoming increasing popular to implement scientific software. 

Author: Suggested change looks good to me.

> Since its inception in 2012, Julia has experienced a steady increase in popularity as practitioners are migrating from scripting languages, such as Python to Julia. Practitioners typically use Julia for large-scale scientific data analysis, e.g., Julia was used in Celeste, a software used in astronomy research. The use of Julia yielded a performance improvement by a factor of 1,000 for Celeste, compared to prior implementation.   

Reviewer: ```suggestion
Since its inception in 2012, Julia has experienced a steady increase in popularity as practitioners are migrating from scripting languages, such as Python.
Practitioners typically use Julia for large-scale scientific data analysis.
For example, Julia was used in [Celeste](https://cs.lbl.gov/news-media/news/2016/celeste-enhancements-create-new-opportunities-in-sky-surveys/), a software used in astronomy research.
Julia yielded a 1000x performance improvement for [Celeste](https://arxiv.org/abs/1611.03404), compared to the prior implementation.   
```

Also, about that 1000x number...is that from changing to Julia alone or also from introducing distributed node parallelism? I don't know Julia...did distributed node parallelism just come for *free* with Julia? If not, it might be worth adding some clarification about that.

Author: Suggested change looks good to me.

The 1000 number is coming from a [blog](https://www.hpcwire.com/off-the-wire/julia-joins-petaflop-club/), which says: "Celeste achieved peak performance of 1.54 petaflops using 1.3 million threads on 9,300 Knights Landing (KNL) nodes of the Cori supercomputer at NERSC – a performance improvement of 1,000x in single-threaded execution." 

Will the phrase "The use of Julia on a supercomputer" make it sound better?

Reviewer: Well, based on that article, it does indeed look like distributed node parallelism comes kinda sorta for free with Julia. So, I am fine leaving that 1000x statement as is.

Author: OK. Sounds good.

But what about bugs in Julia programs? Unmitigated bugs in Julia programs can hinder scientific exploration by generating erroneous results and unexpected failures. Hence, we pursued a research study where we systematically investigated a set of bugs collected from the open source domain. 

### Our Approach 

> We study 742 bugs in total by applying qualitative analysis. Using our qualitative analysis, we derive bug categories and bug symptoms for Julia programs. We also survey 52 practitioners who have contributed to open source Julia programs for further insights. Datasets used in our paper are publicly available [online](https://figshare.com/s/35d775572bb840ebd392).  

Reviewer: ```suggestion
We studied 742 bugs via qualitative analysis. Using our qualitative analysis, we derive bug categories and bug symptoms for Julia programs. We also surveyed 52 practitioners who have contributed to open source Julia programs for further insights. Datasets used in our paper are publicly available [online](https://figshare.com/s/35d775572bb840ebd392).  
```

Author: Suggested change looks good to me.

### Major Findings 

- We identify nine types of bugs 

> - Of the identified nine types of bugs, three have not been reported in other types of software systems, namely, `polyglot`, `pre-compilation`, and `world age'. 

Reviewer: ```suggestion
- Of these, three have not been reported in other types of software systems, namely, `polyglot`, `pre-compilation`, and `world age`. 
```

Author: Suggested change looks good to me.
